src/leaflet.filelayer.js

```diff
@@ -63,46 +63,117 @@
+            if (this._isParameterMissing(file, 'file')) {
+                return;
+            }
+
             // Check file size
-            var fileSize = (file.size / 1024).toFixed(4);
-            if (fileSize > this.options.fileSizeLimit) {
-                this.fire('data:error', {
-                    error: new Error(
-                        'File size exceeds limit (' +
-                        fileSize + ' > ' +
-                        this.options.fileSizeLimit + 'kb)'
-                    )
-                });
+            if (!this._isFileSizeOk(file.size)) {
                 return;
             }
 
-            // Check file extension
-            ext = file.name.split('.').pop();
-            parser = this._parsers[ext];
+            // Get parser for this data type
+            parser = this._getParser(file.name);
             if (!parser) {
-                this.fire('data:error', {
-                    error: new Error('Unsupported file type ' + file.type + '(' + ext + ')')
-                });
                 return;
             }
+
             // Read selected file using HTML5 File API
             reader = new FileReader();
             reader.onload = L.Util.bind(function (e) {
                 var layer;
                 try {
-                    this.fire('data:loading', { filename: file.name, format: ext });
-                    layer = parser.call(this, e.target.result, ext);
-                    this.fire('data:loaded', { layer: layer, filename: file.name, format: ext });
+                    this.fire('data:loading', { filename: file.name, format: parser.ext });
+                    layer = parser.processor.call(this, e.target.result, parser.ext);
+                    this.fire('data:loaded', { layer: layer, filename: file.name, format: parser.ext });
                 } catch (err) {
                     this.fire('data:error', { error: err });
                 }
             }, this);
-            reader.readAsText(file);
-
+            // Testing trick: tests don't pass a real file, but an object with file.testing set to true. 
+            // This object cannot be read by reader, just skip it.
+            if (!file.testing) {
+                reader.readAsText(file);
+            }
             // We return this to ease testing
             return reader;
         },
 
+        loadData: function (data, name, ext) {
+            var parser;
+
+            // Check required parameters
+            if ((this._isParameterMissing(data, "data"))
+              || (this._isParameterMissing(name, "name"))) {
+                return;
+            }
+
+            // Check file size
+            if (!this._isFileSizeOk(data.length)) {
+                return;
+            }
+
+            // Get parser for this data type
+            parser = this._getParser(name, ext);
+            if (!parser) {
+                return;
+            }
+
+            // Process data
+            var layer;
```

Comment:
All `var` should be declared at the top of a function (so with the `parser` var)